crates/claims/crates/data-integrity/core/src/suite.rs

```diff
@@ -0,0 +1,326 @@
+//! Cryptographic suites.
+use iref::Iri;
+use ssi_claims_core::{ProofValidity, Verifiable};
+use ssi_core::Referencable;
+use ssi_crypto::MessageSigner;
+use ssi_json_ld::JsonLdNodeObject;
+use ssi_verification_methods_core::{
+    InvalidVerificationMethod, SignatureError, Signer, VerificationError, VerificationMethod,
+    VerificationMethodResolver,
+};
+use std::convert::Infallible;
+
+use crate::{
+    sign,
+    signing::{self, sign_single},
+    ExpandedConfiguration, ExpandedConfigurationRef, ExpandedType, Proof, ProofConfiguration,
+    ProofConfigurationCastError, ProofConfigurationRefExpansion, ProofRef, Proofs,
+    UnsupportedProofSuite,
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransformError {
+    #[error("Expansion failed: {0}")]
+    ExpansionFailed(String),
+
+    #[error("RDF deserialization failed: {0}")]
+    LinkedData(#[from] linked_data::IntoQuadsError),
+
+    #[error("JSON serialization failed: {0}")]
+    JsonSerialization(json_syntax::SerializeError),
+
+    #[error("expected JSON object")] // TODO merge it with `InvalidData`.
+    ExpectedJsonObject,
+
+    #[error("invalid data")]
+    InvalidData,
+
+    #[error("unsupported input format")]
+    UnsupportedInputFormat,
+
+    #[error("invalid proof options: {0}")]
+    InvalidProofOptions(InvalidOptions),
+
+    #[error("invalid verification method: {0}")]
+    InvalidVerificationMethod(InvalidVerificationMethod),
+
+    #[error("internal error: `{0}`")]
+    Internal(String),
+}
+
+impl From<ProofConfigurationCastError<InvalidVerificationMethod, InvalidOptions>>
+    for TransformError
+{
+    fn from(value: ProofConfigurationCastError<InvalidVerificationMethod, InvalidOptions>) -> Self {
+        match value {
+            ProofConfigurationCastError::VerificationMethod(e) => {
+                Self::InvalidVerificationMethod(e)
+            }
+            ProofConfigurationCastError::Options(e) => Self::InvalidProofOptions(e),
+        }
+    }
+}
+
+impl From<ProofConfigurationCastError<InvalidVerificationMethod, Infallible>> for TransformError {
+    fn from(value: ProofConfigurationCastError<InvalidVerificationMethod, Infallible>) -> Self {
+        match value {
+            ProofConfigurationCastError::VerificationMethod(e) => {
+                Self::InvalidVerificationMethod(e)
+            }
+            ProofConfigurationCastError::Options(_) => unreachable!(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum HashError {
+    #[error("invalid verification method")]
+    InvalidVerificationMethod,
+
+    #[error("message is too long")]
+    TooLong,
+
+    #[error("invalid message: {0}")]
+    InvalidMessage(Box<dyn 'static + std::error::Error>),
```

Comment:
I'll make it a `String`, I agree this much complexity is not required.

Reply:
Sounds good, also just to clarify when I say "they're not actually useful" it's that the crypto libraries purposefully don't return anything helpful to avoid leaking information

Reply:
I've put a `String` instead, and I also "standardized" all the error types so we have a single error type for signing errors, same for signature/proof validation. Also made sure none of those error types have type parameters because that's annoying to deal with.